Bring PaperWM in line with extensions.gnome.org rules - preparing for submission

jtaala opened the pull request:
Combining #557 with this PR.

This PR (finally) brings PaperWM in line with extensions.gnome.org GJS rules & guidelines, which is necessary for submission of PaperWM to extensions.gnome.org.

See issues #205, #435.

Namely, it:

1. Changes PaperWM to use `enable()` instead of `init()`.  `init()` is executed when loaded (e.g. once when logging in).  From [GJS guide/rules](https://gjs.guide/extensions/review-guidelines/review-guidelines.html#rules):
>
> init() is called by GNOME Shell when your extension is loaded, not when it is enabled. Your extension MUST NOT create any objects, connect any signals, add any main loop sources or modify GNOME Shell here.
>
> It is permitted to create static objects and arrays in the global scope or init() (e.g. enumerations or string lists), but constructing class instances is not permitted. Objects intended to be constructed as instances should be created when enable() is called, and destroyed when disable() is called.

2. renames PaperWM UUID to `paperwm@paperwm.github.com`.  This is needed to avoid several issues related to [#489](https://github.com/paperwm/PaperWM/pull/489) (and also since it needs to be somwhere under the control of PaperWM admins).

3. removes the `metadata.json` version parameter.  This is used by extensions.gnome.org to iterate the version number on new releases.

4. simplifies the `install.sh` script!  Auto enabling has not worked for a long (long) time and just confuses users.  The new install script  now checks if  a non-symlinked PaperWM version exists (e.g. one from extensions.gnome.org... when it's available) and if not, applies the symlink with clear instructions to the user on how to enable it.
   - resolves #425. 

5. an installation from extensions.gnome.org means that users won't need to run the `install.sh` script (yay!)... but that means that the `set-recommended-gnome-shell-settings.sh` script wont' be run as well.  Given this, PaperWM now manages these setttings.  It does this by saving/backing-up (to PaperWM's settings/schema) user's settings for `workspaces-only-on-primary`, `edge-tiling`,  and `attach-modal-dialogs`.  
   - we save these settings to PaperWM's settings for safety (in case gnome terminates before PaperWM can restore etc.)
   - PaperWM will save these settings whenever it is enabled, and restore these setting whenever it is disabled;
   - PaperWM clears the backup'd settings after a successful restore, and will only save them on the next enable if they are clear --> this basically ensures that if gnome terminates (crashes etc.), the user's preferred settings won't be lost.
   - given the above, we've now removed the `set-recommended-gnome-shell-settings.sh`;
   - this simplifies the `install.sh` script.

6. Fixes a lot of issues that were stopping us from a successful submission to extensions.gnome.org:  this includes addressing all issues from the first submission (see #557 for the actual submission issues outlined by the extensions.gnome.org):

- [x] **1 DONE**.   Good pickup and call here - was a bit of a mess.  Including a very interesting hard-coding of the `version` var (was written back before Gnome 40).  In any case, the first commit here was fixing that and removing old/deprecated code that was for gnome 3.xx.
- [x] **2 DONE**.  ~~`sessionModes` removed from metadata.json (not needed in PaperWM).  Also improved the `safeCall` logging to only log if called method exists in module.~~ `unlock-dialog` sessionMode is needed!
   - **Justification:**  PaperWM requires session mode `unlock-dialog` to maintain multi-monitor tiling layouts when locking/unlocking Gnome.  Without this, tiled windows can move to different monitors on gnome unlock (which severely impacts/destroys user workflows).
- [x] **3 DONE**.  I'm not sure of the original reason for setting `Extension` like it was doing.  Replaced with a much simpler `var Extension = imports.misc.extensionUtils.getCurrentExtension();`.
- [x] **4 DONE**.  Removed `convenience.js` as it duplicates current Gnome ExtensionUtils and refactored modules that used it .
- [x] **5 DONE**.  Removed unneeded check.
- [x] **6 DONE.**.  did full review of all `mainloop.timeout_add` instances - most were  actually (intended) as once off calls (e.g. destroy after first handler execution) - changed these to destroy after handler function execution.
- [x] **7 DONE**.  Went through and cleaned up `imports.xxx` references (and moving them to top where possible).
- [x] **8 DONE**. We were actually delegating the animation through `utils.tweener`... which was actually using `Widget.ease(...)`.  This was basically because it was easier to do this when we switched to `ease`.  However, this is quite confusing for users/reviews.  I've taken the opportunity now to rename `Tweener.addTween(...)` to `Ease.addEase(...)` etc.
- [x] **9 DONE** (finally!).  This was a very good point in the review.  Polyfilling in functions on global gnome objects isn't the best approach here.  I've now replaced the polyfills with backwards-compatible utility functions which is a cleaner/safer/better approach and should make this much easier to maintain going forward.
- [x] **10 DONE**.  Replaced the use of `print(...)` across PaperWM. 
- [x] **11 DONE**.  Moved global scope creation of instances to `enable` in `Kludges.sj`, `liveAltTab.js`, `prefs.js`, `tiling.js`.  Also cleaned up  and removed unneeded imports dependencies and improved formatting.
- [x] **12 DONE**. Hmm - we don't see an issue with importing Gtk and using accelerator_parse etc.  But we have cleaned up and removed references to Gtk where not needed.
- [x] **13 DONE**.  We've removed the hot corner overriding/disabling from PaperWM.  This was largely superfluous since hot corner works fine in PaperWM, and if users want to disable it they can do so via the usual gnome setting.  Having two functions that disable hot corner (one in gnome and one in PaperWM) seems superfluous.
- [x] **14 DONE**.  Removed unneeded import in `minimap.js`. 
- [x] **15 DONE**.  `destroy` should not call `dismissDispatcher` since `dismissDispatcher` calls `destroy`... leading to an infinite loop (recursion).
- [x] **16 DONE**.  All `log(...)` lines reviewed and removed superfluous logging; 
- [x] **17 DONE**. Lower-cased PaperWM schema id values and recompiled schema.
- [x] **18 DONE**.  Refactored to  NOT call `setSchemas()` on Settings `init` (is now called on `enable`).
- [x] **19 DONE**.  nulled out vars (that we missed).  Good pickup! 
- [x] **20 DONE**. Now untracking on `destroy()`.
- [x]  **21 DONE**.  Refactored StackOverlay class - added proper destroy method and now removePreview on destroy.
- [x] **22 DONE**. Timeouts are now oneshot (remove/destroy timeout after function callback).
- [x] **23 DONE**.  Added missing semicolons and other cleanup in `virtTiling.js`.
- [x] **24 DONE**. Now using `add_style_class_name` and `remove_style_class_name` to not run over other extension styles.
- [x] **25 DONE**. Removed Gdk import from Tiling.js.
- [x] **26 DONE**.  Removed module self reference from `keybindings.js` and `tiling.js`. 
- [x] **27 DONE**.  Replaced `Gdk.Keymap` (which isn't available in Gnome 40+) with `Clutter.get_default_backend().get_default_seat().get_keymap()` and hence removed `Gdk` import from `keybindings.js`.  Haven't been able to remove `Gtk` since we need `accelerator_parse` from `Gtk`.


### NOTE: given we've changed the UUID, users will likely need to rerun the `./install.sh` script and login/logout after checking this branch out.

jtaala committed:
Moved away from `init()` methods to align with requirements for gnome
extensions review process: https://gjs.guide/extensions/review-guidelines/review-guidelines.html#basics

jtaala committed:
Clean-up of unneeded file.

jtaala committed:
Need to keep oldSpaces, oldMonitors maps for restoring layout when
locking/unlocking.

jtaala committed:
Rename UUID to align with https://gjs.guide/extensions/review-guidelines/review-guidelines.html#metadata-json-must-be-well-formed

jtaala committed:
FIX: prefs should be init() (which is executed when opening
settings/prefs).

jtaala committed:
Removed "version" from metadata.json, as directed in https://gjs.guide/extensions/review-guidelines/review-guidelines.html#metadata-json-must-be-well-formed

jtaala committed:
FIX: prefs/settings now work if no metadata.version is defined.

jtaala committed:
Added a `generate-extension-zip.sh` script that generates a zip with
(only) the necessary files Gnome needs for PaperWM.  Designed for
generating an artifact to submit to extensions.gnome.org.

jtaala committed:
Added message re changing the UUID to README.md.

jtaala committed:
Removed old/unused polyfills (for versions prior to 40).

jtaala committed:
FIX: removed obscure version array (which forced version[0] to be "3"...
i.e. assume was gnome 3.x.

jtaala committed:
[2] DONE. Removed sessionModes (not needed) and improved `safecall`
logging (only log if method exist in module.

jtaala committed:
FIX: forgot about topbar init() - migrated to enable().

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

jtaala committed:
[5] DONE.  Removed unneeded check (for gnome 3.x).

jtaala committed:
[3] DONE.  Replaced with standard `var Extension = imports.misc.extensionUtils.getCurrentExtension();
` insstead.

jtaala committed:
[4] DONE.  Convenience removed as it now duplicates current gnome
ExtensionUtils.

jtaala committed:
[7] DONE. Cleaned up imports.xxx references.

jtaala committed:
Replaced the use of `print(...)` throughout PaperWM.

jtaala committed:
[14] DONE.  Removed unneeded import.

jtaala committed:
[25][27] DONE.  Removed Gdk.Keymap (which isn't in available in Gnome
40+) and removed Gdk dependency from Keybindings.js.

jtaala committed:
[26] DONE. Removed self (module) reference from `keybindings.js` and
`tiling.js`.

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

jtaala committed:
[23] DONE.  Added missing semicolons and other cleanup in
`virtTiling.js`.  Added eslint files to .gitignore.

jtaala committed:
[11] DONE.  Moved global scrope creation of instances to `enable` in
`Kludges.sj`, `liveAltTab.js`, `prefs.js`, `tiling.js`.  Also cleaned up
and removed unneeded imports dependencies and improved formatting.

jtaala committed:
[9] Replaced `raise` and `raise_top` Actor polyfills with utils function.
Removed Clutter.Actor polyfill for `set` and polyfill for `global.display.set_cursor` (checked Gnome 42, 43, & 44 and these are present so polyfill not needed).
Replaced Clutter.Actor polyfill for `reparent` with reparent utils function.
Replaced Cluter.Vertex polyfill with Graphene.Point3D direclty (polyfill
was using this anyway and Graphene.Point3D is available in 42, 43, 44).
Replace Meta.later_add polyfill with backwards compatible `later_add`
utils function.

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

jtaala committed:
[6] DONE. did full review of all mainloop.timeout_add instances - most were
actually (intended) as once off calls - changed these to destroy after
call (return false in calling handler function:
(https://gjs-docs.gnome.org/gjs/mainloop.md#mainloop-timeout_add).

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

jtaala committed:
[8] the name `Tweener` used previously was just a legacy name, we're actually using Widget.ease here.  This was renamed to avoid confusion with the deprecated `Tweener` module.

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

Lythenas commented:
Is there a meaningful difference between startup and enable? I saw that startup is always(?) just called in enable. If that is the case we should maybe just put all of it in the enable function, or give startup a more meaningful name (or split up the enable functions a bit).

jtaala commented:
> Is there a meaningful difference between startup and enable? I saw that startup is always(?) just called in enable. If that is the case we should maybe just put all of it in the enable function, or give startup a more meaningful name (or split up the enable functions a bit).

Hey @Lythenas!

No, there's no meaningful difference there - for some of them I just found it easier to rename the original `init` to `startup` and call it in the enable (well, I actually originally called it something like `preenable`).

Including all in enable() would be okay, except in `kludges.js` - that would make `enable` a very large function.  Have any ideas for a more meaningful name?

jtaala committed:
FIX: typo in virtTiling import (St).

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

Lythenas commented on kludges.js:
```diff
@@ -527,7 +331,7 @@ function enable() {
     // Fullscreen windows aren't special in paperWM and might not even be
     // visible, so hiding notifications makes no sense.
-    with (imports.ui.messageTray) {
+    with (MessageTray) {
```
Also what does this `with` do? I have never seen that in JS. According to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with it should be avoided.

jtaala replied:
This is a good question!  I'm not sure - I just changed the imports reference.  Let's look at changing it.

jtaala replied:
So! turns out this monkey patching of MessageTray is actually not needed at all (should have testing just removing this for #553).  That is, default gnome behaviour now actually does what #553 was doing by modifying the monkey patch!

Removed it entirely.  Problem solved!

Lythenas commented on kludges.js:
```diff
@@ -388,16 +202,22 @@ function restoreMethod(obj, name) {
+var settings;
+var wmSettings;
+function startup() {
+    settings = ExtensionUtils.getSettings();
+    wmSettings = new Gio.Settings({schema_id: 'org.gnome.desktop.wm.preferences'});
```
Unfortunately I can't attach a comment to an unchanged line... So here everything in one comment. I would:

- Put all `registerOverridePrototype` calls into `enableOverride` (and maybe rename it to `setupOverrides`.
- Put all the swipe tracker related stuff in a function `setupSwipeTrackers`
- Put these lines in `enable` and also the line `signals = new utils.Signals();`

And we could similarly split up the `enable` function. E.g. into:

- `setupSignals`
- `monkeyPatchWindowManager`
- `monkeyPatchWorkspace`
- `monkeyPatchMessageTray`

(I don't know what modifying the prototype in JS is usually called, but "monkey patching" is often used in Python for similar things)

jtaala replied:
Hey @Lythenas!  I've implemented these (check the latest).  Simplified quite a bit and implemented the override functions into one place.

Also removed MessageTray as not needed/used now.

Cheers!

jtaala committed:
Removed MessageTray monkey patching - turns out not needed since default
gnome behaviour now aligns with #553 (what 553 was doing on the monkey
path... gnome does now).

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

jtaala committed:
[18] DONE.  Now not calling setSchemas() on Settings init - only did
that to ensure prefs.sj (extensions prefs) had those init'd when opening
prefs.  Now just calling setSchemas() direclty from prefs to init those
vars when needed (e.g. when opening prefs).

jtaala committed:
Merge branch 'submission-to-extensions.gnome.org' into address-issues-from-submission-1-to-extensions.gnome.org

jtaala committed:
Moved prototype overriding functions to register (where they should be).
Clean-up MessageTray (not need anymore).

jtaala committed:
Clean up of kludges overrides and restores.  Refactor/simplification of
enable methods.

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

jtaala committed:
Missed one override function (integrated into overrides registry).

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

jtaala committed:
[16] DONE. All `log(...)` lines reviewed and remove superfluous logging.

jtaala committed:
[19] DONE.  Nulled out vars that we missed.

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

jtaala committed:
Cleanup keybindings.js.  Replaced `accelerator_parse_with_keycode` with
`accelerator_parse`.

jtaala committed:
[17] DONE. Lowercased or.gnome... fields in schemas.

jtaala committed:
[21] WIP.

jtaala committed:
Refactor of StayOverlay remove preview. Added `destroy()` function.

jtaala committed:
[22] DONE. Timeouts are now oneshot (remove/destroy timeout after function callback).

jtaala committed:
[24] DONE. Now using `add_style_class_name` and `remove_style_class_name` to not
run over other extension styles.

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

jtaala committed:
Moved `initUserConfig` to `enableUserConfig` - now loads user.css styles
without needing a login/logout.  Also checks for user config on enable.

jtaala committed:
Renamed last `startup` to `setupActions` to be more descriptive.

jtaala committed:
Manage runtimeDisables (e.g. hotcorners etc.) better (and more
declaratively).

jtaala committed:
Merge branch 'submission-to-extensions.gnome.org' into address-issues-from-submission-1-to-extensions.gnome.org

jtaala committed:
Added `runtimeDisables` which saves the orignal setting for `attach-modal-dialogs`,
`workspaces-only-on-primary`, and `edge-tiling`, disables them, and then
restores the original setting on PaperWM disable.  Needed since with an
extensions installed from extensions.gnome.org we can't use the install
script to set recommended settings.

jtaala committed:
Removed hotcorner overriding.  Hot corner works fine in PaperWM.  If
users want to disable hotcorne then they should do so with the usual
method.  Having two functions that disable hot corners (one in gnome and
one in PaperWM seems superfluous).

jtaala committed:
Rewrite of runtimeDisables.  Now we save a backup of the user's setting to PaperWM settings (schema) for safety (in case gnome terminates etc.).  This ensures original user settings will be restored on next PaperWM disable.

jtaala committed:
Merge branch 'address-issues-from-submission-1-to-extensions.gnome.org' into submission-to-extensions.gnome.org

jtaala committed:
set-recommended-gnome-shell-settings.sh is not needed now, since PaperWM
now saves user settings for `workspaces-only-on-primary`, `edge-tiling`
and `attach-modal-dialogs` to PaperWM settings on enable, and restores
these whenever PaperWM is disabled.  We save these settings to PaperWM's
settings for safety (in case gnome terminates before PaperWM can
restore etc.) - this ensures we persistently save these settings.

jtaala commented:
@Lythenas!

I'm keen on getting this merged in.  It's a big one though, and has lots of changes, improvements, and bugfixes.  It's a hell of alot cleaner, meaner, and does things in a much more gnome "friendly" way.

I've been testing it solidly this past week (and as I've been developing this PR).

Of note, please see point 4 & 5:

> 4. simplifies the `install.sh` script!  Auto enabling has not worked for a long (long) time and just confuses users.  The new install script  now checks if  a non-symlinked PaperWM version exists (e.g. one from extensions.gnome.org... when it's available) and if not, applies the symlink with clear instructions to the user on how to enable it.
> 
> 5. an installation from extensions.gnome.org means that users won't need to run the `install.sh` script (yay!)... but that means that the `set-recommended-gnome-shell-settings.sh` script wont' be run as well.  Given this, PaperWM now manages these setttings.  It does this by saving/backing-up (to PaperWM's settings/schema) user's settings for `workspaces-only-on-primary`, `edge-tiling`,  and `attach-modal-dialogs`.  
>    - we save these settings to PaperWM's settings for safety (in case gnome terminates before PaperWM can restore etc.)
>    - PaperWM will save these settings whenever it is enabled, and restore these setting whenever it is disabled;
>    - PaperWM clears the backup'd settings after a successful restore, and will only save them on the next enable if they are clear --> this basically ensures that if gnome terminates (crashes etc.), the user's preferred settings won't be lost.
>    - given the above, we've now removed the `set-recommended-gnome-shell-settings.sh`;
>    - this simplifies the `install.sh` script.

jtaala committed:
Updated README.md re PaperWM managing the recommended gnome settings.

jtaala committed:
Large simplification of the `install.sh` script.  Auto enabling hasn't
worked for a long time and just confuses users.  The script now check if
a non-symlinked version exists (e.g. one from extensions.gnome.org) and
if not, apply the symlink with directions to the user on how to enable
it.

Lythenas commented:
@jtaala I ran into the first error. After installing and enabling it for the first time:

```
PaperWM
Error occured in user @enable:

Tiling is undefined

enable@/home/ms/.config/paperwm/user.js:29:5
safeCall@/home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/extension.js:61:52
run@/home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/extension.js:49:17
enable@/home/ms/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/extension.js:97:12
_callExtensionEnable@resource:///org/gnome/shell/ui/extensionSystem.js:196:38
async*_onEnabledExtensionsChanged@resource:///org/gnome/shell/ui/extensionSystem.js:552:24
_loadExtensions/<@resource:///org/gnome/shell/ui/extensionSystem.js:627:18
createCheckedMethod/<@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:730:46
enableExtension@resource:///org/gnome/shell/ui/extensionSystem.js:223:29
EnableExtension@resource:///org/gnome/shell/ui/shellDBus.js:447:38
_handleMethodCall@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:329:38
_wrapJSObject/<@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:408:34
```

It still works fine on my first monitor but PaperWM on my second monitor is broken. (Gnome 44.2 if that matters)

The same error also reappears after locking and unlocking.

After logging out completely and logging back in the error still appears but it seems like the second monitor also works correctly now.

Note that I did a complete fresh install, removed old symlink, made a new clone, checked out and ran the install script.

jtaala commented:
> @jtaala I ran into the first error. After installing and enabling it for the first time:

Thanks @Lythenas. 

That error is actually from your `user.js` (user defined code) in your `home/ms/.config/paperwm/user.js` - looks like you have some custom code in there (it's not the standard tempate `user.js` that PaperWM installs).

Can you post your `user.js` content?

A couple of options - you can fix (or delete) line 29 in user.js - or a better option, is I'll wrap the user.js `init` and `enable` calls in try/catch, since we can't know/control what users do there (but at least can try harden PaperWM against user defined init/enable issues and print them to log).

jtaala commented:
Oh, actually it's already wrapped so that error is just a log output saying that there's an issue in the users own `user.js`.

I'm interested though in the second monitor PaperWM being broken though.  Can you elaborate there?  I'll do some more second monitor testing.

jtaala commented:
> The same error also reappears after locking and unlocking.
> 
> After logging out completely and logging back in the error still appears but it seems like the second monitor also works correctly now.
> 
> Note that I did a complete fresh install, removed old symlink, made a new clone, checked out and ran the install script.

Okay, just checking, on first install (of this branch) did you logout/login and the second monitor PaperWM wasn't working right?  Or were you on X11 and just did a `alt+f2` `r`?  After your login/logout are you still seeing funniness?  can you try reproduce that?

I'm running second monitor now, haven't seen anything yet (apart from the focus mode icon being placed funny after and lock/unlock, until you move your mouse to that monitor and it corrects itself - I'll fix that though).

Oh, and def post your `user.js` - we can then see what the issue is - if it's doing nothing useful then you can just blow it away to get rid of that error notification in the logs.

jtaala committed:
FIX/REVERT: PaperWM needs `session-modes` `unlock-dialog` to maintain
multi-monitor tiling layouts when locking/unlocking gnome.  Without this
windows can move to different monitors on unlock.

jtaala commented:
@Lythenas, all, please pull the latest here - I've restored session mode `unlock-dialog` in metadata.json.

> - [x] **2 DONE**.  ~~`sessionModes` removed from metadata.json (not needed in PaperWM).  Also improved the `safeCall` logging to only log if called method exists in module.~~ `unlock-dialog` sessionMode is needed!
>   - **Justification:**  PaperWM requires session mode `unlock-dialog` to maintain multi-monitor tiling layouts when locking/unlocking Gnome.  Without this, tiled windows can move to different monitors on Gnome unlock (which severely impacts user workflows).

jtaala commented:
@Lythenas, scrap that - found it.  So, since PaperWM doesn't run stuff in `init` anymore, it wasn't calling `user.js`'s `init` method - that's why you `user.js` (which has some stuff in `enable` that relies on `init` being run) fell over.

I'll fix this up, so any previous `user.js` code runs as before.

jtaala committed:
FIXES: extension.js now runs `init` for modules (not currently used in
PaperWM modules but can be used in `user.js` modules).  Updated
`example/user.js` file to reflect new approaches for init/enable
(previous `user.js` files will still run fine).

jtaala committed:
Improved handling of user config files and config dir creation.  Also
better handles cases where `user.js` doesn't exist (i.e. make it
optional).

jtaala committed:
Moved `user.js` template to a better place (`config` folder).

Lythenas commented:
Also on gnome 42.5 I run into this:

```
-En PaperWM
Error occured in tiling @enable:

global.compositor is undefined

later_add@/home/mase271683/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/utils.js:549:9
updateMonitor@/home/mase271683/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/topbar.js:777:15
finish@/home/mase271683/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/tiling.js:1698:20
monitorsChanged@/home/mase271683/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/tiling.js:1778:9
Spaces@/home/mase271683/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/tiling.js:1604:14
enable@/home/mase271683/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/tiling.js:2723:14
safeCall@/home/mase271683/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/extension.js:62:52
run@/home/mase271683/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/extension.js:45:14
enable@/home/mase271683/.local/share/gnome-shell/extensions/paperwm@paperwm.github.com/extension.js:107:12
_callExtensionEnable@resource:///org/gnome/shell/ui/extensionSystem.js:183:32
_onEnabledExtensionsChanged/<@resource:///org/gnome/shell/ui/extensionSystem.js:532:35
_onEnabledExtensionsChanged@resource:///org/gnome/shell/ui/extensionSystem.js:532:14
createCheckedMethod/<@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:533:46
enableExtension@resource:///org/gnome/shell/ui/extensionSystem.js:210:29
EnableExtension@resource:///org/gnome/shell/ui/shellDBus.js:448:38
_handleMethodCall@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:310:38
_wrapJSObject/<@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:387:34

```

jtaala committed:
FIX: for gnome 42, 43, global.compositor.get_laters check.

jtaala commented:
> Also on gnome 42.5 I run into this:
> 
> ```
> -En PaperWM
> Error occured in tiling @enable:
> 
> global.compositor is undefined
> ```

Thanks @Lythenas.  Fixed!

jtaala committed:
Formatting cleanup of tiling and topbar.

jtaala committed:
FIX: focusmode icon on enable with multi-monitor position was messed-up.

jtaala committed:
FIXES #558.  We now manage grabbed (drag & drop) windows in
ClickOverlay.  If a window is grabbed-dragged-dropped on a different
monitor, then execute ClickOverlay.select() (which deactivates
ClickOverlay and immediately selects/activates that window).

Lythenas commented:
> @Lythenas, scrap that - found it. So, since PaperWM doesn't run stuff in `init` anymore, it wasn't calling `user.js`'s `init` method - that's why you `user.js` (which has some stuff in `enable` that relies on `init` being run) fell over.
> 
> I'll fix this up, so any previous `user.js` code runs as before.

Ah, sorry I didn't realize the error was in my user.js, sorry. But the error is fixed with you latest changes, thanks.

For the record, I think my user.js is just the template one with some `Tiling.defwinprop`s from before it was configurable in the settings UI.

I will try your other fixes on gnome 42 tomorrow.

Lythenas commented:
Also should we maybe update the `user.js` template to not use `init`? But I would prefer to still support calling it to not break other users (although I don't know how many really have much code there, since it is a pain to edit this on wayland).

jtaala committed:
Updated README.md re being able to relaod `user.css` by enable/disable
PaperWM.  Updated to copy latest `metadata.json` on enable.

jtaala commented:
> Also should we maybe update the `user.js` template to not use `init`? But I would prefer to still support calling it to not break other users (although I don't know how many really have much code there, since it is a pain to edit this on wayland).

Yep, already have.  Using `init` is still valid for several use cases, so I think it's fine for `user.js` template - it's just that PaperWM modules don't need/rely on `init` anymore (and there's some advantages to doing so).  The template is also now simplified and cleaned up, with some modules already defined at top level.

For example, you can now reload `user.css` styles by disabling/enable PaperWM (without needing to restart).  I updated the README.md re this.  Actually, I'll re-add the `Main.loadTheme()` method reference in README.md (since that's still valid).

jtaala committed:
Re-added in alternative way of reloading `user.css` in README.md.

jtaala committed:
FIX #545.  Gdk.KEY_less ("<") causes issues when trying to bind in
PaperWM.  Safer to use 'comma'. Replace "<" with comma when binding.

Lythenas commented:
Tested on Gnome 42 today and it also seems to work fine.

Lythenas commented on README.md:
````diff
@@ -25,18 +28,16 @@ Clone the repo and check out the branch supporting the Gnome Shell version you'r
 ./install.sh # install, load and enable paperwm
 ```
-> ⚠️ You will likely need to restart Gnome shell after installing PaperWM, e.g. logout then login, or restart in place with an `alt-F2` and entering `r` (X11 only).
+> #### ➡️ You may need to restart Gnome shell after installing PaperWM, e.g. logout then login, or restart in place with an `alt-F2` and entering `r` (X11 only).
````
```suggestion
> #### ➡️ You will need to restart Gnome shell after installing PaperWM, e.g. logout then login, or restart in place with an `alt-F2` and entering `r` (X11 only).
```

I feel like this would be better because wayland is now more or less the default.

Lythenas commented on README.md:
```diff
@@ -25,18 +28,16 @@ Clone the repo and check out the branch supporting the Gnome Shell version you'r
-`$XDG_DATA_HOME/gnome-shell/extensions/`. It will then ask if you want to apply
-the recommended settings (see [Recommended
-Settings](#recommended-gnome-shell-settings)) and lastly it will ask to enable PaperWM.
+`$XDG_DATA_HOME/gnome-shell/extensions/`. It will then ask if you want to enable PaperWM.
```
```suggestion
`$XDG_DATA_HOME/gnome-shell/extensions/` (usually in `~/.local/share/gnome-shell-extension/`). It will then ask if you want to enable PaperWM.
```

or simply directly write the path instead of `$XDG_DATA_HOME`

Lythenas commented on config/user.js:
```diff
@@ -0,0 +1,16 @@
+var Tiling = Extension.imports.tiling;
+var Keybindings = Extension.imports.keybindings;
+
+function init() {
+    // Runs _only_ once on Gnome login (or Gnome shell restart)
+}
```
Should we maybe just remove this from the template, since it is not recommended to be used anyway?

If you want to keep it we should at least make it clear that all code should usually be in `enable`.

jtaala replied:
@Lythenas - yeah, I'm starting to think this too.  Even more, [the latest review by extensions.gnome.org](https://extensions.gnome.org/review/43111) want use to remove that init() entirely  - so `user.js` init wouldn't be called.

Now, if they have code in user.js init(), it won't break PaperWM but it will give an error notification to the user.  I'm not sure how many use that, but it would require moving to enable/disable (which is better here anyways).

We could remove the above init(), and maybe use the template with some comments re that.

Lythenas replied:
Unfortunately I also can't tell you how many people even have code in `user.js`.

Maybe to (kind of) retain the old behavior, we could call the init of `user.js` on the very first call of `enable` (right before `enable` of the `user.js`). I expect this wouldn't break anything for the user.

Also I think it would also be reasonable for use to just remove the support for `user.js` entirely. It is not actually that much less code than creating a second extension (basically just the `metadata.json` is missing). In that second extension we (or users) would not be restricted by gnome guidelines. (We can of course still provide a template for that and/or a script to create that second extension)

jtaala replied:
So, I pushed some changes to:

- remvoe `init()` from user.js template;
- remove `init()` from extensions.js;
- run user.js `init` on first PaperWM enable;
- also only run `warnAboutGnomeShellVersionCompatibility` on first enable.

Lythenas commented on extension.js:
```diff
@@ -77,31 +78,17 @@ var SESSIONID = ""+(new Date().getTime());
-var Extension, convenience;
+/**
+ * Runs once on extension init().
+ * Not run on PaperWM modules, but if a user as a `user.js` module defined
```
```suggestion
 * Not run on PaperWM modules, but if a user has a `user.js` module defined
```

Lythenas commented on extension.js:
```diff
@@ -111,8 +98,16 @@ function enable() {
-    if (run('enable'))
+    SESSIONID += "#";
+    Extension = imports.misc.extensionUtils.getCurrentExtension();
+    warnAboutGnomeShellVersionCompatibility();
```
Since we are now calling this on every enable, it might "spam" the user a bit with warnings. We should add a check to only run it once.

jtaala replied:
Oh good point - yeah, we should actually use in functions!

jtaala replied:
Yes, I think so.  The other option is remove this check / warning entirely.  `metadata.json` now is stricter in versions supported, e.g. if it's not running `42`, `43`, or `44` you won't be able enable this extension.

The only advantage I can see with keeping this is if we enable early/prelim support for `45` - although, lessons from `44` issues suggest we should just do that in a `45` branch for users that are on `45` to test etc.

I'm thinking we don't need this anymore - thoughts?

jtaala replied:
So, I pushed some changes to: only run `warnAboutGnomeShellVersionCompatibility` on first enable (and not subsequent enable/disable calls).

Did this as an interim until we decide if we want to keep this check (or if you can think of other use/corner cases where this check woudl be needed).

Lythenas replied:
Yes, I think now with the new `release` branch we can also remove this warning. Because now people will not as easily accidentally update paperwm to an experimental state.

Lythenas commented on generate-extension-zip.sh:
```diff
@@ -0,0 +1,14 @@
+
+# Creates a zip of (only) the necessary files required by Gnome for the PaperWM extension.
+# Designed for submitting a zip to extensions.gnome.org.
+/usr/bin/zip -r paperwm@paperwm.github.com.zip \
```
```suggestion
zip -r paperwm@paperwm.github.com.zip \
```

Why are you using an absolute path?

Lythenas commented on install.sh:
```diff
@@ -1,85 +1,36 @@
-    echo "Could not enable PaperWM automatically. You may need to enable it manually."
-    printf 'Debug output: %s\n' "$unescaped"
-fi
+/usr/bin/gnome-extensions enable ${UUID}
```
```suggestion
gnome-extensions enable ${UUID}
```

Lythenas commented on keybindings.js:
```diff
@@ -1,51 +1,47 @@
 var display = global.display;
 
-var KEYBINDINGS_KEY = 'org.gnome.Shell.Extensions.PaperWM.Keybindings';
+var KEYBINDINGS_KEY = 'org.gnome.shell.extensions.paperwm.keybindings';
```
Can we actually use this constant string in the functions below?

jtaala replied:
Oh good point - yeah, we should actually use in functions!

Lythenas commented on keybindings.js:
```diff
@@ -254,23 +242,10 @@ function byId(mutterId) {
-            return actionHandler(mw, Tiling.spaces.selectedSpace, {display, binding});
-        };
-}
+var asKeyHandler = (actionHandler) =>
+    (display, mw, binding) => {
+        return actionHandler(mw, Tiling.spaces.selectedSpace, { display, binding });
+    };
```
I think there is no point keeping this an arrow function since it is on the global scope anyway.

Lythenas commented on tiling.js:
```diff
@@ -1325,13 +1281,26 @@ border-radius: ${borderWidth}px;
         }
     }
 
+    /**
+ * Updates workspace topbar icon positions.
+ */
```
```suggestion
    /**
     * Updates workspace topbar icon positions.
     */
```

Lythenas commented on utils.js:
```diff
@@ -448,26 +429,31 @@ var Signals = class Signals extends Map {
+/**
+ * Note the name `Tweener` used previously was just a legacy name, we're actually using
+ * Widget.ease here.  This was renamed to avoid confusion with the deprecated `Tweener`
+ * moddule.
```
```suggestion
 * module.
```

jtaala replied:
~~@Lythenas - I like these changes.  Can you please do a PR (into this PR) with your suggested changes?~~

~~Will make it much easier for me.~~

jtaala replied:
@Lythenas  - I've added most of these changes.  Can talk re the other points.

Lythenas reviewed (changes requested):
I took some time to try to review this more thoroughly. But it is possible I still missed something. Github (or git) sometimes does not display the diff that great.

Most of the things I found are just nitpicks. So after fixing them I'm fine with approving this.

jtaala committed:
Implemented changes requested by @lythenas.

jtaala committed:
Removed init() from extensions.js.  Refactored user.js safecalls to call
init on first enable (to support legacy user.js's).

jtaala committed:
Improved user config setup, and only add config to search path if it
isn't already in searchpath.

jtaala committed:
Removed warnAboutGnomeShellVersionCompatibility method - not needed now
as metadata.json versioning is more strict (42, 43, 44).

jtaala committed:
Update generate-extension-zip.sh

Co-authored-by: Matthias Seiffert <matthias-seiffert@hotmail.de>
